File: Boj-Kotlin&Java/백준2512번예산.java
package 백준;

import java.io.BufferedReader;
import java.io.IOException;
import java.io.InputStreamReader;

public class 백준2512번예산 {
    static int n,m;
    static int[] arr;
    static int right=-1,left=0;
    public static void main(String[] args) throws IOException {
        BufferedReader br=new BufferedReader(new InputStreamReader(System.in));
        n=Integer.parseInt(br.readLine());
        arr=new int[n];
        String[] input=br.readLine().split(" ");
        for(int i=0; i<n; i++) {
            arr[i] = Integer.parseInt(input[i]);
            right=Math.max(right,arr[i]);
        }
        m=Integer.parseInt(br.readLine());
        long answer=0,sum=0;
        while (left<=right){
            sum=0;
            int mid=(left+right)/2;
            for(int i=0; i<n; i++)
                sum+=Math.min(arr[i],mid);
            if(sum<=m){
                answer=mid;
                left=mid+1;
            }else
                right=mid-1;
        }
        System.out.println(answer);
    }
}

File: Boj-Kotlin&Java/백준 2512번 예산.kt
package 백준

import kotlin.math.min

private var m:Long=0
fun main() {
    val br=System.`in`.bufferedReader()
    val n=br.readLine().toInt()
    val arr=br.readLine().split(" ").map{it.toInt()}
    m=br.readLine().toLong()
    var left=0
    var right=arr.max()
    var answer=0
    var sum:Long=0
    while (left<=right!!){
        var mid=(left+right)/2
        sum=0
        for(i in arr.indices)
            sum+= min(arr[i],mid)
        if(sum<=m){
            answer=mid
           left=mid+1
        }else{
            right=mid-1
        }
    }
    println(answer)
}